Store pointer encoding configuration in boolbv_widtht

tautschnig opened the pull request:
Rather than handling the object/offset bit width in multiple places,
make boolbv_widtht the only place to provide this information.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs reviewed (approved):
Makes sense to centralise this functionality.

tautschnig committed:
bv_endianness_mapt is specific to bv_pointerst

The special-casing of the endianness encoding handled by
bv_endianness_mapt is really specific to bv_pointerst, not boolbvt in
general. Thus, move this data structure to bv_pointers.cpp and enable
overriding the endianness map from boolbvt.

This commit is refactoring only, no changes in behaviour intended.

tautschnig committed:
Enable overriding of boolbvt::boolbv_width

With upcoming changes to pointer encoding it will be necessary to enable
children of boolbvt to define their own bit encoding width of data
types.

This commit is refactoring only, no changes in behaviour intended.

tautschnig committed:
Make pointer flattened width configurable by bv_pointerst

In preparation of tweaking the propositional encoding, place all knowledge of
the bit width of encoded pointers in bv_pointerst. Also, do not hard-code the
width just once, but really look at each type.
